### ISSUE TITLE
`not-callable` false positive in type narrowing

### ISSUE THREAD

**Issue by mttbernardini** [pylint-dev/pylint#7412]
### Bug description

Pylint emits a false positive when calling `func()`, which shouldn't happen because of the type narrowing of the `if`.

```py
from typing import Optional

class W():
    def __init__(self, f):
        self.f = f

    def __call__(self, *args, **kwargs):
        print("> Call")
        r = self.f(*args, **kwargs)
        print("< Ret")
        return r

func: Optional[W] = None

def test():
    if func is not None:
        func()
 ```

I noticed that if I add to the previous code the following snippet, the error disappears:

```py
def init():
    global func
    func = W(lambda x: x-2)
 ```

Similarly, the error is not emitted if `func` is initialized directly in global scope.

However I think this kind of inference is wrong because the above code could be a module and `func` could be set from outside and the `if` is guarding that scenario 🤔 

### Command used

```shell
pylint --disable=C,R callable.py
```

### Pylint output

```shell
************* Module callable
callable.py:17:8: E1102: func is not callable (not-callable)
```

### Expected behavior

`pylint` shouldn't report `E1102` in this case.

### Pylint version

```shell
pylint 2.15.0
astroid 2.12.5
Python 3.8.2 (default, Jun  8 2021, 11:59:35) 
[Clang 12.0.5 (clang-1205.0.22.11)]
```

### OS / Environment

MacOS 11.5.2

**Comment by DanielNoord** [comment]
This is just something that `pylint` is not good at. We struggle with some control flow inference using `if` and we don't take type annotation into account currently. This could change in the future though.

**Comment by joelspadin** [comment]
I found a similar case which is likely the same root cause:

```py
class Base:
    pass

class Version1(Base):
    pass

class Version2(Base):
    pase

def get_type(version: int) -> type[Base]:
    if version == 1:
        return Version1
    if version == 2:
        return Version2
    raise NotImplementedError()

version = 1
cls = get_type(version)
obj = cls() # not-callable
```

If I change `get_type` to
```py
def get_type(version: int) -> type[Base]:
    if version == 1:
        return Version1
    if version == 2:
        return Version2
    raise NotImplementedError()
    return Version1
```
then it no longer gives the false positive `not-callable` but instead (correctly) gives a `unreachable` on the extra return I added.

**Comment by kojiromike** [comment]
Is this the same as #6005?

**Comment by Pierre-Sassoulas** [comment]
Right, thank you for triaging @kojiromike, it's very similar, but this one seems more generic and the case given by @joelspadin shed light on another limitation.

**Comment by jacobtylerwalls** [comment]
Fixed in https://github.com/pylint-dev/astroid/pull/1189 (astroid 2.13)
